pulsar-client-cpp/lib/ConsumerImpl.cc

```diff
@@ -1326,8 +1336,26 @@ void ConsumerImpl::getLastMessageIdAsync(BrokerGetLastMessageIdCallback callback
             callback(ResultUnsupportedVersionError, MessageId());
         }
     } else {
-        LOG_ERROR(getName() << " Client Connection not ready for Consumer");
-        callback(ResultNotConnected, MessageId());
+        TimeDuration next = std::min(remainTime, backoff->next());
+        if (next.total_milliseconds() <= 0) {
+            LOG_ERROR(getName() << " Client Connection not ready for Consumer");
+            callback(ResultNotConnected, MessageId());
+            return;
+        }
+        remainTime -= next;
+
+        timer->expires_from_now(next);
+
+        timer->async_wait([this, backoff, remainTime, timer, next,
+                           callback](const boost::system::error_code& ec) -> void {
```

Comment:
It's better to capture `shared_from_this()` here because when the callback is executed, the reference count of `std::shared_ptr<ConsumerImpl>` might be zero so that `this` points to an invalid address. Capturing the `shared_from_this()` can extend the lifetime of the object to ensure `this` always points to a valid address.

Reply:
Looks like there is a similar problem in the current codes. https://github.com/apache/pulsar/blob/c4bd7ae15a1e37075fb5a0b3b4636591ed2462c8/pulsar-client-cpp/lib/AckGroupingTrackerEnabled.cc#L153-L158

Could you guide me on how to capture `shared_from_this()`? I haven't found a good way to solve it.

Reply:
Your reference is an example. `self` is what `shared_from_this()` returns, this lambda captures `self` by value so that the reference count will increase by one. After the lambda ends, the reference count will decrease by one, so the object keeps valid (refcnt is always > 0) during the lifetime of the lambda.

Reply:
Oh. I get it. Thanks.